### PR TITLE
Auto accept binary changes for Upgraded properties

### PR DESCRIPTION
**Pull request by asodja**
With that we don't need to update `accepted-public-api-changes.json` anymore after upgrade. We handle only properties upgraded with `@UpgradedProperty`. All others have to be handled manually.

Additionally we also check, that all interceptod methods are actually present in previous binary: e.g. one could upgrade a property, that has only a getter: there is no need to try to intercept a setter (although we currently don't support that case with `@UpgradedProperty`, but we could).

Fixes https://github.com/gradle/gradle/issues/26293

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/upgrades/UpgradedProperty.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    private final String methodDescriptor;
+    private final String containingType;
+    private final List<UpgradedMethod> upgradedMethods;
+
+    public UpgradedProperty(String containingType, String propertyName, String methodName, String methodDescriptor, List<UpgradedMethod> upgradedMethods) {
+        this.containingType = containingType;
+        this.propertyName = propertyName;
+        this.methodName = methodName;
+        this.methodDescriptor = methodDescriptor;
+        this.upgradedMethods = upgradedMethods;
```
Comment (lptr) [inline build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/upgrades/UpgradedProperty.java:33]: 💅 This should make a defensive copy with `ImmutableList.copyOf()`.

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/upgrades/UpgradedProperty.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+        @Override
+        public String toString() {
+            return "UpgradedMethod{" +
+                "name='" + name + '\'' +
+                ", descriptor='" + descriptor + '\'' +
+                '}';
+        }
+    }
+}
```
Comment (lptr) [inline build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/upgrades/UpgradedProperty.java:92]: 💭 Had we had Java 17, all this could be expressed in like 10 lines with two `record`s... :(

File: subprojects/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java [hunk 3]
```diff
@@ -110,6 +111,8 @@ public Collection<Result> readRequest(ExecutableElement method) {
         if (projectName == null) {
             // We validate project name here because we want to fail only if there is an @UpgradedProperty annotation used in the project
             return Collections.singletonList(new InvalidRequest("Project name is not specified or is empty. Use -A" + PROJECT_NAME_OPTIONS + "=<projectName> compiler option to set the project name."));
+        } else if (!method.getParameters().isEmpty() || !method.getSimpleName().toString().startsWith("get")) {
+            return Collections.singletonList(new InvalidRequest(String.format("Method '%s.%s' annotated with @UpgradedProperty should be a simple getter: name should start with 'get' and method should not have any parameters.", method.getEnclosingElement(), method)));
```
Comment (lptr) [inline subprojects/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java:115]: 🌮 Nice!

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy [hunk 4]
```diff
@@ -41,7 +43,9 @@ class BinaryCompatibilityHelper {
         FileCollection sourceRoots,
         String currentVersion,
         File apiChangesJsonFile,
-        Directory projectRootDir
+        Directory projectRootDir,
+        File currentUpgradedPropertiesFile,
+        File baseUpgradedPropertiesFile
```
[the remark this answers is not in the document]
Comment (asodja) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy]: Fixed that 🎉

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/upgrades/UpgradedProperties.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        Map<String, UpgradedProperty> currentMethods = context.getUserData(CURRENT_METHODS_OF_UPGRADED_PROPERTIES);
+        Map<String, UpgradedProperty> oldMethods = context.getUserData(OLD_METHODS_OF_UPGRADED_PROPERTIES);
+
+        if (jApiMethod.getCompatibilityChanges().contains(METHOD_REMOVED)) {
+            return isOldSetterOfUpgradedProperty(jApiMethod, oldMethods) || isOldBooleanGetterOfUpgradedProperty(jApiMethod, oldMethods);
+        } else if (jApiMethod.getCompatibilityChanges().contains(METHOD_RETURN_TYPE_CHANGED)) {
+            return isCurrentGetterOfUpgradedProperty(jApiMethod, currentMethods) && isOldGetterOfUpgradedProperty(jApiMethod, oldMethods);
+        } else if (jApiMethod.getCompatibilityChanges().contains(METHOD_ADDED_TO_PUBLIC_CLASS)) {
+            return isCurrentGetterOfUpgradedProperty(jApiMethod, currentMethods);
+        }
```
[the remark this answers is not in the document]
Comment (asodja) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/upgrades/UpgradedProperties.java:90]: Good point, added some comments

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/ExtractGradleApiInfoTask.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.Collections;
+
+/**
+ * Extracts Gradle API information from the given Gradle distribution archives.
+ */
+@CacheableTask
+public abstract class ExtractGradleApiInfoTask extends DefaultTask {
```
[the remark this answers is not in the document]
Comment (asodja) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/ExtractGradleApiInfoTask.java:45]: Good point. After some thoughts I figured we do a bit more than just a transform, we also some filtering here and in the end we output just a file for a specific jar. With transform we would probably need to do the filtering that at configuration time, and for tests we would need to duplicate some part of the code (or abstract it to some method).

Due to that I decided that I would keep it in a task.

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild.binary-compatibility.gradle [hunk 7]
```diff
@@ -104,34 +110,40 @@ dependencies {
...
-    def baseVersion = moduleIdentity.version.map { it.baseVersion.version }
     def isSnapshot = moduleIdentity.snapshot
     inputs.property('acceptedViolations', acceptedViolations.toAcceptedChangesMap())
     inputs.property("baseline.version", compatibilityBaselineVersion)
     inputs.property("currentVersion", baseVersion)
     def apiSourceFolders = configurations.currentSources.incoming.artifactView { lenient(true) }.files
     inputs.files("apiSourceFolders", apiSourceFolders)
-    def currentClasspath = configurations.currentApiClasspath.incoming.artifactView { lenient(true) }.files
     inputs.files(currentClasspath)
+    inputs.files(extractGradleApiInfo)
```
[the remark this answers is not in the document]
Comment (asodja) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild.binary-compatibility.gradle:138]: I have some logic in the `ExtractGradleApiInfoTask`. But other logic looks very specific to `JapicmpTask` and uses Japicmp types. So I didn't extract it out of it.

[the remark this answers is not in the document]
Comment (asodja) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild.binary-compatibility.gradle:138]: Or were you thinking about all other logic that is not related to my change?

[the remark this answers is not in the document]
Comment (lptr) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild.binary-compatibility.gradle:138]: I mean the current state, which includes old and new changes. It's a `JapicmpTask` but we are adding a lot of extra functionality. I wonder if we could either create our own task from the ground up, or extend `JapicmpTask` via a proper task type.

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/ExtractGradleApiInfoTask.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (lptr) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/ExtractGradleApiInfoTask.java:45]: I think it's fine to do this as a task, too, but I don't see the things you brought up as reasons not to convert to transforms anyway. Still, task should be fine.

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/rules/UpgradePropertiesRuleSetup.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    private static Map<MethodKey, UpgradedProperty> mapOldMethodsOfUpgradedProperties(List<UpgradedProperty> upgradedProperties) {
+        return upgradedProperties.stream()
+            .flatMap(UpgradePropertiesRuleSetup::mapOldMethodsOfUpgradedProperty)
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private static Stream<Map.Entry<MethodKey, UpgradedProperty>> mapOldMethodsOfUpgradedProperty(UpgradedProperty upgradedProperty) {
+        return upgradedProperty.getUpgradedMethods().stream().map(upgradedMethod -> {
```
Comment (lptr) [inline build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/rules/UpgradePropertiesRuleSetup.java:74]: 💅 `.map()` should be on a new line.

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/upgrades/UpgradedProperties.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        Map<String, UpgradedProperty> oldMethods = context.getUserData(OLD_METHODS_OF_UPGRADED_PROPERTIES);
+        String descriptor = jApiMethod.getOldMethod().get().getSignature();
+        String containingType = jApiMethod.getjApiClass().getFullyQualifiedName();
+        String key = getMethodKey(containingType, name, descriptor);
+        return oldMethods.containsKey(key) ? Optional.of(key) : Optional.empty();
+    }
+
+    public static String getMethodKey(String containingType, String methodName, String descriptor) {
+        return containingType + "#" + methodName + descriptor;
+    }
```
[the remark this answers is not in the document]
Comment (lptr) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/upgrades/UpgradedProperties.java]: IIUC, you started using `MethodKey` from `:model-core`. This is an internal API reference; if we change `MethodKey`, it will mess up the Gradleception build. I don't think it's a likely thing, but it would be good to avoid adding more internal references.

Also, we should at some point introduce checks to prevent more internal API references from our own plugins and from our build logic; that validation will flag this, too.

[the remark this answers is not in the document]
Comment (asodja) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/upgrades/UpgradedProperties.java]: This is not `MethodKey` from `:model-core` but it's my own implementation in build logic. I will rename the class, to avoid confusion.

[the remark this answers is not in the document]
Comment (asodja) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/upgrades/UpgradedProperties.java]: Fixed

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/rules/UpgradePropertiesRuleSetup.java (hunk 8 above)
Comment (asodja) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/rules/UpgradePropertiesRuleSetup.java:74]: Fixed

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild.binary-compatibility.gradle (hunk 7 above)
[the remark this answers is not in the document]
Comment (asodja) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild.binary-compatibility.gradle:138]: To describe current state:
So currently it works like that: we registere our onw own implementation of violation rules and some pre/post processors. Due to that there is a lot of logic (and inputs) in one task. We actually have our own implementation  of JapicmpTask that then delegates work via work action to [japicmp-gradle-plugin](https://github.com/melix/japicmp-gradle-plugin).

So I guess we would need to create our own task from the ground up to fix that.

[the remark this answers is not in the document]
Comment (lptr) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild.binary-compatibility.gradle:138]: I see. I think we are fine for the PR here.

In the context of evolving the task API we should think about how we imagine users implementing similar solutions without extending tasks etc. What building blocks, best practices can we offer? Not something to answer here, but we should come up with one.

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/upgrades/UpgradedProperties.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (lptr) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/upgrades/UpgradedProperties.java]: Ah, sorry, it's my mistake, should have checked out your branch.

**Comment by asodja** [comment]
@bot-gradle test and merge
